src/urllib3/multipart/decoder.py

```diff
@@ -0,0 +1,149 @@
+    """
+
+    def __init__(self, content: bytes, encoding: str):
+        #: Encoding used for the body part to decode body and headers
+        self.encoding = encoding
+        headers: dict[str, str] = {}
+        # Split into header section (if any) and the content
+        headerbytes, separator, bodybytes = content.partition(b"\r\n\r\n")
+        if b"\r\n\r\n" != separator:
+            raise ImproperBodyPartContentError("content does not contain CR-LF-CR-LF")
+
+        #: The bytes containing the body of this part
+        self.content = bodybytes
+        if headerbytes != b"":
+            headers.update(_header_parser(headerbytes.lstrip(), encoding))
+        #: The headers associated with this part
+        self.headers = _collections.HTTPHeaderDict(headers)
+
+    @property
+    def text(self) -> str:
+        """Content of the ``BodyPart`` in unicode."""
+        return self.content.decode(self.encoding)
+
+
+MD = typing.TypeVar("MD", bound="MultipartDecoder")
+
+
+class MultipartDecoder:
+    """This parses the full multipart/form-data payload.
+
+    The ``MultipartDecoder`` object parses the multipart payload of
+    a bytestring into a tuple of ``Response``-like ``BodyPart`` objects.
+
+    The basic usage is::
+
+        import requests
+        from requests_toolbelt import MultipartDecoder
+
+        response = request.get(url)
+        decoder = MultipartDecoder.from_response(response)
+        for part in decoder.parts:
+            print(part.headers['content-type'])
+
+    If the multipart content is not from a response, basic usage is::
+
+        from requests_toolbelt import MultipartDecoder
+
+        decoder = MultipartDecoder(content, content_type)
+        for part in decoder.parts:
+            print(part.headers['content-type'])
+
+    For both these usages, there is an optional ``encoding`` parameter. This is
+    a string, which is the name of the unicode codec to use (default is
+    ``'utf-8'``).
+    """
+
+    def __init__(self, content: bytes, content_type: str, encoding: str = "utf-8"):
+        #: Original Content-Type header
+        self.content_type = content_type
+        #: Response body encoding
+        self.encoding = encoding
+        #: Parsed parts of the multipart response body
+        self.parts: tuple[BodyPart, ...] = tuple()
+        self._find_boundary()
+        self._parse_body(content)
+
+    def _find_boundary(self) -> None:
+        ct_info = tuple(x.strip() for x in self.content_type.split(";"))
+        mimetype = ct_info[0]
+        if mimetype.split("/")[0].lower() != "multipart":
+            raise NonMultipartContentTypeError(
+                f"Unexpected mimetype in content-type: '{mimetype}'"
+            )
+        for item in ct_info[1:]:
+            attr, _, value = item.partition("=")
+            if attr.lower() == "boundary":
+                self.boundary = encode_with(value.strip('"'), self.encoding)
```

Comment:
I wonder what happens if you receive a value which isn't properly quoted – e.g. if `value` starts with an open quote but there's no matching close quote.